### ISSUE TITLE
populating ComboBox model after startup shows first item selected but does not execute `selected` callback

### ISSUE THREAD

**Issue by Be-ing** [slint-ui/slint#8611]
### Bug Description

In my application, I have a ComboBox whose model cannot be populated immediately when Slint starts because its values come from GStreamer asynchronously. A message bus from GStreamer sends messages when new camera devices are available, which is monitored by a Future with `slint::spawn_local`.

In practice, GStreamer sends the messages over its bus for the available cameras almost instantaneously after startup, but not before `run` is called on the Slint window. The result is that the Slint ComboBox shows that the first item of the model is selected. However, the `selected` callback of the ComboBox doesn't get triggered for this first item. This leaves the application in a confusing, invalid state, where it looks like something should have happened because of what is shown in the ComboBox, but to make it actually happen, the user has to click the ComboBox and re-select the first item (or select a different item).

Alternatively, Slint could show nothing in the ComboBox until the user explicitly selects an item for the first time. This would be less confusing, but cumbersome for users unless the application developer programatically selects the ComobBox's first item when populating the model.

### Reproducible Code (if applicable)

```rust
    let device_monitor = gst::DeviceMonitor::new();
    let caps = gst::Caps::new_empty_simple("video/x-raw");
    device_monitor.add_filter(Some("Video/Source"), Some(&caps));
    slint::spawn_local({
        let mut messages = device_monitor.bus().stream();
        async move {
            while let Some(msg) = messages.next().await {
                use gst::MessageView;
                // Determine whether we want to quit: on EOS or error message
                // we quit, otherwise simply continue.
                match msg.view() {
                    MessageView::Eos(..) => break,
                    MessageView::Error(err) => {
                        error!(
                            "Error from {:?}: {} ({:?})",
                            err.src().map(|s| s.path_string()),
                            err.error(),
                            err.debug()
                        );
                        break;
                    }
                    MessageView::DeviceAdded(d) => {
                        let name = SharedString::from(d.device().display_name().as_str());
                        let None = camera_map.borrow().get(&name) else {
                            warn!(
                                "GStreamer DeviceMonitor added device {name}, which already existed"
                            );
                            break;
                        };
                        info!("Device added: {name}");
                        camera_map.borrow_mut().insert(name.clone(), d.device());
                        camera_model.push(name);
                    }
                    _ => debug!("Message received"),
                }
            }
        }
    })
    .unwrap();
```

https://codeberg.org/Be.ing/dance-video-recorder/src/commit/5da08f106db0e6f3bfc8686606f9fd01d848be40/src/main.rs#L138

commit with workaround: https://codeberg.org/Be.ing/dance-video-recorder/commit/07ce5dd14dd7d5302b96b25732acfafc3aa3ca41

### Environment Details

- Slint Version: 1.11.0
- Platform/OS: Linux (Wayland)
- Programming Language: Rust
- Backend/Renderer: winit

### Product Impact

This is an inconvenience that can be worked around with just a few lines of code when populating the model:

```rust
                        if camera_model.row_count() == 1 {
                            app.upgrade().unwrap().invoke_camera_selected(name);
                        }
```

However, I don't think this workaround should be needed. Slint shouldn't put the application in a state where it shows an item has been selected but hasn't executed the `selected` callback.

**Comment by tronical** [comment]
Do I understand correctly that *after* populating the combobox, you'd like the selected callback to also be invoked automatically, as if the user selected the first time?

**Comment by tronical** [comment]
My feeling is that when an application shows the a combobox to the user but the data to be shown it it isn't ready yet (because it's not available), then combo box should be disabled (by the developer).

When the data becomes enabled, it should also be enabled. In the absence of an explicit choice, I suppose the current item of the ComboBox becomes either the first or last item of the model. My feeling is that at this point, the application code should act as if the user selected whatever the application decides should be the first item.

I think the `selected` callback should only be invoked if the user selected an entry.

**Comment by Be-ing** [comment]
> the application code should act as if the user selected whatever the application decides should be the first item.

The issue is that Slint is currently implicitly showing the first item of the model without the application doing anything but adding data to the model. If Slint left the ComboBox in a blank state with no item selected, it would make sense for the application to have explicitly select the first (or last, or whatever is appropriate for the application) item. I would expect the application programatically selecting an item would implicitly trigger the `selected` callback as well, as if the user selected it.

It would be convenient for my use case if Slint automatically showed the first item and executed the `selected` callback for it. Though now that I think about this more, there might be use cases where the application puts the items in some specific order and programatically selects some item that isn't the first (for example, loading stored preferences), in which case automatically executing the `selected` callback for the first item could potentially have undesirable effects (for example, perhaps a brief flicker of colors if the ComboBox is controlling the colors of something on screen).

**Comment by tronical** [comment]
> The issue is that Slint is currently implicitly showing the first item of the model without the application doing anything but adding data to the model. If Slint left the ComboBox in a blank state with no item selected, it would make sense for the application to have explicitly select the first (or last, or whatever is appropriate for the application) item. 

A blank state is also not very useful for users - arguably a bug.

> there might be use cases where the application puts the items in some specific order and programatically selects some item that isn't the first (for example, loading stored preferences), in which case automatically executing the selected callback for the first item could potentially have undesirable effects (for example, perhaps a brief flicker of colors if the ComboBox is controlling the colors of something on screen)

Exactly what I was thinking, too.

It really feels like `selected` should only be invoked when the user changes the current item (could also be via accessibility, fwiw). But I'm starting to wonder if this callback is all that useful when it may lead to applications relying on it and getting into the weird state.

The way it looks like to me - and I may very well be missing something here - is that if the combobox selection is static, `selected` is straight-forward. But if it's dynamic (or lazily populated), `changed current-index => { ... }` seems more reliable?

**Comment by Be-ing** [comment]
> A blank state is also not very useful for users - arguably a bug.

Indeed, it's not very useful for users, but it is useful to developers as an indication that "hey, you haven't selected anything, you should probably do that".

**Comment by Be-ing** [comment]
> The way it looks like to me - and I may very well be missing something here - is that if the combobox selection is static, selected is straight-forward. But if it's dynamic (or lazily populated), changed current-index => { ... } seems more reliable?

I tested switching from the `selected` callback to `changed current-index`:

https://codeberg.org/Be.ing/dance-video-recorder/commit/3c67b6d334481d925bfef85469e0e3feb1fec1ee

That has the same issue though. The first item of the model is shown in the ComboBox without the `changed current-index` callback getting called, leaving the application in a confusing state.

Curiously, `changed current-value` doesn't have this issue: https://codeberg.org/Be.ing/dance-video-recorder/commit/ec1a3a128b679ae3b4f8cde63d6fc48703900ca2 The first item is selected and the callback is executed.

**Comment by Be-ing** [comment]
> But I'm starting to wonder if this callback is all that useful when it may lead to applications relying on it and getting into the weird state.

I was not aware of the `changed attribute` feature until you mentioned it here because I haven't written code with Slint in a while. I presume the `selected` callback of ComboBox predates that feature. I think it would be reasonable to deprecate it in favor of `changed current-index` or `changed current-value`... though similar callbacks exist on other standard widgets and it'd be odd to change this for for just one of them.

`changed property` requires more boilerplate:

```
callback camera-selected(string);

CameraSelector := ComboBox {
    changed current-value => { camera-selected(self.current-value); }
}
```

It would be nice if there was a shorthand that didn't require writing out the function like
```
callback camera-selected(string) <=> changed CameraSelector.current-value
CameraSelector := ComboBox {}
```

**Comment by tronical** [comment]
> That has the same issue though. The first item of the model is shown in the ComboBox without the `changed current-index` callback getting called, leaving the application in a confusing state.
> 
> Curiously, `changed current-value` doesn't have this issue: https://codeberg.org/Be.ing/dance-video-recorder/commit/ec1a3a128b679ae3b4f8cde63d6fc48703900ca2 The first item is selected and the callback is executed.

I guess because the model you're populating lazily still has a valid index zero, while however the row changes?

**Comment by tronical** [comment]
>  I think it would be reasonable to deprecate it in favor of changed current-index or changed current-value... though similar callbacks exist on other standard widgets and it'd be odd to change this for for just one of them.

Plus `selected` makes sense if you *are* interested in reacting to user input, not user input *and* programmatic changes.

**Comment by tronical** [comment]
It's unclear to me what the actionable thing here is, TBH, so I'm closing this. `selected` is documented to be invoked on user input, invoking it also on programmatic changes seems like a breaking change to me.

With a pre-populated ComboBox all is well, AFAICS. With a combobox where the underlying model changes dynamically, the application can't just rely on `selected` indeed. It's true that when just replacing say the model, a `changed current-index` isn't enough either, but it's unclear what the generic API should be for what may be perhaps best solved on the application level: When changing the model, adapt related state and make an explicitly choice what entry the user should see in the combo box (blank state issue).

**Comment by Be-ing** [comment]
At the very least the documentation could use improvement.

> selected is documented to be invoked on user input, invoking it also on programmatic changes seems like a breaking change to me.

This is what the documentation says now:

"A value was selected from the combo box. The argument is the currently selected value."

This language does not distinguish between user input and programmatic changes.

**Comment by tronical** [comment]
Fair point, I'll clarify the wording :)
